nix/eval-machine-info.nix

```diff
@@ -11,104 +11,170 @@
 let
   call = x: if builtins.isFunction x then x args else x;
 
-  # Copied from nixpkgs to avoid <nixpkgs> import
-  optional = cond: elem: if cond then [elem] else [];
-
   zipAttrs = set: builtins.listToAttrs (
     map (name: { inherit name; value = builtins.catAttrs name set; }) (builtins.concatMap builtins.attrNames set));
 
-  flakeExpr = (builtins.getFlake flakeUri).outputs.nixopsConfigurations.default;
-
-  networks =
-    let
-      getNetworkFromExpr = networkExpr:
-        (call (import networkExpr)) // { _file = networkExpr; };
+  flake = builtins.getFlake flakeUri;
+  flakeExpr = (builtins.tryEval flake.outputs.nixopsConfigurations.default).value or { };
 
-      exprToKey = key: { key = toString key; };
+  nixpkgsBoot = toString <nixpkgs> ; # this will be replaced on install by nixops' nixpkgs input
+  libBoot = import "${nixpkgsBoot}/lib";
 
-      networkExprClosure = builtins.genericClosure {
-        startSet = map exprToKey networkExprs;
-        operator = { key }: map exprToKey ((getNetworkFromExpr key).require or []);
+  baseMods = lib: [
+    {
+      options.nixpkgs = lib.mkOption {
+        type = lib.types.path;
+        description = "Path to the nixpkgs instance used to buld the machines.";
+        defaultText = lib.literalDocBook "The 'nixpkgs' input to either the provided flake or nixops' own.";
+        default = (builtins.tryEval flake.inputs.nixpkgs).value or nixpkgsBoot;
       };
-    in
-      map ({ key }: getNetworkFromExpr key) networkExprClosure
-      ++ optional (flakeUri != null)
-        ((call flakeExpr) // { _file = "<${flakeUri}>"; });
+      config._module.freeformType = with lib.types;attrsOf anything;
+    }
+    flakeExpr
+  ] ++ networkExprs;
 
-  network = zipAttrs networks;
+  evalBoot = libBoot.evalModules {
+    specialArgs = args;
+    modules = baseMods libBoot;
+  };
 
-  evalConfig =
-    if flakeUri != null
-    then
-      if network ? nixpkgs
-      then (builtins.head (network.nixpkgs)).lib.nixosSystem
-      else throw "NixOps network must have a 'nixpkgs' attribute"
-    else import (pkgs.path + "/nixos/lib/eval-config.nix");
+  inherit (evalBoot.config) nixpkgs;
 
-  pkgs = if flakeUri != null
-    then
-      if network ? nixpkgs
-      then (builtins.head network.nixpkgs).legacyPackages.${system}
-      else throw "NixOps network must have a 'nixpkgs' attribute"
-    else (builtins.head (network.network)).nixpkgs or (import <nixpkgs> { inherit system; });
+  pkgs = nixpkgs.legacyPackages.${system} or (import nixpkgs { inherit system; });
+  lib = nixpkgs.lib or pkgs.lib or libBoot;
+  inherit (lib) mkOption types;
+  inherit (builtins) removeAttrs;
 
-  inherit (pkgs) lib;
+in
+rec {
+  inherit nixpkgs;
 
-  # Expose path to imported nixpkgs (currently only used to find version suffix)
-  nixpkgs = builtins.unsafeDiscardStringContext pkgs.path;
+  net = lib.evalModules {
+    specialArgs = args;
+    modules = baseMods lib ++ [
```

Comment:
We'll probably want a more specific `freeformType`. `anything` was acceptable for bootstrapping because we don't use the value, but for legacy `nodes` we need to inject `pkgs`.